### PR TITLE
shellenv: Add zsh site-functions to `fpath`

### PR DESCRIPTION
**Pull request by HeroCC**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

This PR adds the brew zsh site-functions directory to the fpath during `brew shellenv`, allowing us to enable completions in one line instead of the seven currently suggested on the [Shell Completions](https://docs.brew.sh/Shell-Completion) page.

The shell-completions documentations page states we don't do this by default because managing completions is complex. This is true, however, I still think it's fair to pick the low-hanging-fruit on behalf of the user; many installations already call `compinit` on their behalf, and by including the homebrew fpath we can give those users completions for free. Remaining users only need to call compinit (assuming they have shellenv already setup). 

If this gets accepted, I can make another PR to update the docs, suggesting users eval the shellenv command instead of the manual fpath manipulation instructions.

For the implementation itself, I made a few design choices for backwards compatibility (documented in the commit description), but I'm flexible if someone prefers it another way.

### CODE REVIEW

**Comment by MikeMcQuaid** [comment]
I'm not sure this should be enabled by default. I have seen non-zero numbers of Homebrew packages with issues and people who wish to not have default ZSH behaviours overridden like this.

**Comment by HeroCC** [comment]
Hmm, maybe we could append to fpath instead of prepend to give priority to other functions? I see `fpath` as serving a very similar function to `PATH` / `MANPATH` -- when you install packages to brew, I'd expect all of it's resources to be available. So I found myself surprised to find this not installed by default, like all of the other `*PATH`s. 

> I have seen non-zero numbers of Homebrew packages with issues

Are these issues with completions, or autoloaded functions in general? Any functions / completions installed here would still need to be actively `autoload`ed before ZSH adopts them, so I don't think this would be a significant behavior change for users that don't already opt-in to completions or autoload their functions. 

Skimming through a [search](https://github.com/search?q=repo%3AHomebrew%2Fhomebrew-core+%28completion.install+OR+site-functions%29&type=code) of `homebrew-core` it seems like the vast majority of references to site-functions is for completions anyway.

**Comment by MikeMcQuaid** [comment]
@Homebrew/maintainers any opinions either way on whether we make `brew shellenv` do this?

**Comment by ZhongRuoyu** [comment]
To me this looks fine per se, as it does simplify configuration of Zsh completions.

Although the rest in the [docs](https://docs.brew.sh/Shell-Completion#configuring-completions-in-zsh) still hold true (i.e., additional configuration is still needed), I doubt if we can do anything better than this.

**Review by ZhongRuoyu — APPROVED** [review]

**Comment by MikeMcQuaid** [comment]
> To me this looks fine per se, as it does simplify configuration of Zsh completions.
> 
> Although the rest in the [docs](https://docs.brew.sh/Shell-Completion#configuring-completions-in-zsh) still hold true (i.e., additional configuration is still needed), I doubt if we can do anything better than this.

What things are/are not required if we do this?

Would be good to update the docs in this PR, too, so this is more easily understood.

**Review by MikeMcQuaid — CHANGES_REQUESTED** [review]
Thanks for the PR, still want a bit more discussion before merge!

**Comment by ZhongRuoyu** [comment]
> What things are/are not required if we do this?

For a vanilla Zsh setup:

```zsh
if type brew &>/dev/null
then
  # Not required; replaced by the changes in this PR
  FPATH="$(brew --prefix)/share/zsh/site-functions:${FPATH}"

  # Below are still required, and must be placed after `brew shellenv`
  autoload -Uz compinit
  compinit
fi
```

For Oh My Zsh:

> Note that if you are using Oh My Zsh, it will call `compinit` for you when you source `oh-my-zsh.sh`. In this case, instead of the above, add the following line to your `~/.zshrc`, before you source `oh-my-zsh.sh`:
> ```zsh
> FPATH="$(brew --prefix)/share/zsh/site-functions:${FPATH}"
> ```

Most of that is still true, except that it's `brew shellenv` that needs to be placed before `source oh-my-zsh.sh`.

**Comment by ZhongRuoyu** [comment]
> Is this something we could/should make not required? Presumably without this it's a harmless no-op.

I updated my earlier comment so hopefully it's clearer now!

**Comment by ZhongRuoyu** [comment]
> Anything similar we could/should do for Bash (in another PR)?

I haven't played around with Bash completions so I'm not quite sure about this one.

**Comment by MikeMcQuaid** [comment]
@ZhongRuoyu Thanks, that all makes sense. I'm 👍🏻 on adding this to `brew shellenv` in that case but would like the documentation to be updated in this PR before merge.

**Comment by ZhongRuoyu** [comment]
Pushed an documentation update.

One thing I didn't realise earlier is that `brew shellenv` lives in `~/.zprofile` so the ordering is already implicitly enforced. (My earlier comment has been updated accordingly.) But now, since `brew shellenv` is in `~/.zprofile`, it's also sourced by non-interactive shells, so then it's probably not idea to do `compinit` with `brew shellenv` 😅.

**Review by MikeMcQuaid — APPROVED** [review]
Thanks so much for your first contribution! Without people like you submitting PRs we couldn't run this project. You rock, @HeroCC!

Also thanks to @ZhongRuoyu for updating the docs here.

**Comment by ZhongRuoyu** [comment]
We do have `eval "$(brew shellenv)"` in `~/.zshrc` on Linux, so ordering is a problem on Linux. 446cda2080d16f6dd1e0c9c7ac73ca22a6835efd covers that. Sorry for the back and forth!

**Review by MikeMcQuaid — APPROVED** [review]

**Pull request merged**
